src/timer.rs

```diff
@@ -0,0 +1,85 @@
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use std::thread;
+use std::time::{Duration, Instant};
+
+/// Milliseconds since ANCHOR.
+static RECENT: AtomicU64 = AtomicU64::new(0);
+lazy_static! {
+    static ref ANCHOR: Instant = Instant::now();
+}
+
+/// Convert a duration to millisecond.
+#[inline]
+pub fn duration_to_millis(dur: Duration) -> u64 {
+    dur.as_secs() * 1000 + dur.subsec_millis() as u64
+}
+
+/// Returns milliseconds since ANCHOR.
+///
+/// ANCHOR is some fixed point in history.
+pub fn now_millis() -> u64 {
+    let res = Instant::now();
+    let mut t = duration_to_millis(res.saturating_duration_since(*ANCHOR));
+    let mut recent = RECENT.load(Ordering::Relaxed);
+    loop {
+        if recent > t {
+            t = recent;
+            break;
+        }
+        match RECENT.compare_exchange_weak(recent, t, Ordering::Relaxed, Ordering::Relaxed) {
+            Ok(_) => break,
+            Err(r) => recent = r,
+        }
+    }
+    RECENT.store(t, Ordering::Relaxed);
```

Comment:
seems this store is redundant either for the first or second break condition

Reply:
Good catch. The line is forgotten to removed, it's written before the loop.